File: graph/find_string.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++
class Solution {
  string s;
  vector<string> vs;
  int N;
  vector<int> marked;
  int K;

  void solve(int i, int k) {
    if (i == N) {
      vs.push_back(s);
      return;
    }
    for (int j = 0; j < k; ++j) {
      s.push_back(j + '0');
      solve(i + 1, k);
      s.pop_back();
    }
  }

  bool dfs(string &t, int k) {
    if (k == vs.size())
      return true;

    for (int j = 0; j < vs.size(); ++j) {
      if (!marked[j]) {
        if (neighbour(t.substr(t.size() - N), vs[j])) {
          t += vs[j][N - 1];
          marked[j] = 1;
          if (dfs(t, k + 1))
            return true;
          marked[j] = 0;
          t.pop_back();
        }
      }
    }
    return false;
  }

  bool neighbour(const string &s1, const string &s2) {
    for (int i = 1; i < N; ++i) {
      if (s1[i] != s2[i - 1])
        return false;
    }
    return true;
  }

public:
  string findString(int n, int k) {
    // code here
    N = n;
    solve(0, k);
    marked.assign(vs.size(), 0);
    K = k;
    string t(N, '0');
    marked[0] = 1;
    dfs(t, 1);

    return t;
  }
};

//{ Driver Code Starts.
int main() {
  int t;
  cin >> t;
  while (t--) {
    int N, K;
    cin >> N >> K;
    Solution ob;
    string ans = ob.findString(N, K);
    int ok = 1;
    for (auto i : ans) {
      if (i < '0' || i > K - 1 + '0')
        ok = 0;
    }
    if (!ok) {
      cout << -1 << endl;
      continue;
    }
    unordered_set<string> st;
    for (int i = 0; i + N - 1 < ans.size(); i++) {
      st.insert(ans.substr(i, N));
    }
    int tot = 1;
    for (int i = 1; i <= N; i++)
      tot *= K;
    if (st.size() == tot) {
      cout << ans.size() << endl;
    } else {
      cout << -1 << endl;
    }
  }
  return 0;
}
// } Driver Code Ends

File: recursion/find_string.py
# User function Template for python3


class Solution:
    def findString(self, N, K):
        # code here
        if N == 1:
            return "".join(str(c) for c in range(K))
        s = set()

        def f(k, curr):
            if k == N:
                s.add(curr)
                return
            for i in range(K):
                curr = curr + str(i)
                f(k + 1, curr)
                curr = curr[:-1]

        curr = ""
        f(0, curr)

        self.ans = "0" * N
        s.remove(self.ans)

        def dfs(k):
            if k == K**N:
                return True
            for i in range(K):
                t = self.ans[-(N - 1) :] + str(i)
                if t in s:
                    s.remove(t)
                    self.ans += str(i)
                    if dfs(k + 1):
                        return True
                    self.ans = self.ans[0:-1]
                    s.add(t)

            return False

        dfs(1)
        return self.ans


# {
# Driver Code Starts
# Initial Template for Python 3

import sys

sys.setrecursionlimit(10**6)

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        N, K = map(int, input().split())

        ob = Solution()
        ans = ob.findString(N, K)
        ok = 1
        for i in ans:
            if ord(i) < 48 or ord(i) > K - 1 + 48:
                ok = 0
        if not ok:
            print(-1)
            continue
        d = dict()
        i = 0
        while i + N - 1 < len(ans):
            d[ans[i : i + N]] = 1
            i += 1
        tot = pow(K, N)
        if len(d) == tot:
            print(len(ans))
        else:
            print(-1)
# } Driver Code Ends
